### ISSUE TITLE
Ordered observe with projection behaves differently on the client and server side

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#3597]
I made this [test package](https://github.com/mitar/minimongo-diff), which runs Minimongo tests on the server. It exposes the issue with ordered observe with projection. Behavior is not the same with Minimongo and MongoDB collection.

**Comment by glasser** [comment]
I'm going to do my best to at least analyze if not fix this in the next few weeks as part of https://trello.com/c/xR0t2vRd/82-livequery-mongo-driver-and-minimongo-maintenance

**Comment by glasser** [comment]
Are you referring to all the failures of the form `fail — assert_equal - expected ["changed",{"a":3},0,{"a":2}] - not`?

That's just what I referred to in #2315.  In minimongo, there's a (problematic) guarantee that observe callbacks run synchronously with the insert/update/remove call, which means that you can write tests of observe without needed to use write fences (or the equivalent) to know when to check for observe callbacks.

So this isn't really a bug — minimongo provides a stronger guarantee than server-side observe here, and the tests rely on that.

**Comment by mitar** [comment]
The issue is that this is prevent one to write the same both on the client and server side, which is one of the Meteor's main powers, no? So something should be changed. Or Minimongo or server side. But the same code should run in the same way on both.

Just imagine how hard is to debug this things in real code. Where for some reason order of things is different on the client and server side. Here we have nice tests which show you examples of issues and you will just disregard this?

**Comment by glasser** [comment]
I'm not disregarding it, just saying that it's a duplicate of #2315.

**Comment by mitar** [comment]
OK, then I will complain again only after #2315 is fixed if the tests will be still failing. ;-) Thanks for all your work!

**Comment by mitar** [comment]
I updated the [example](https://github.com/mitar/minimongo-diff) with latest Meteor code. There are still differences on the server side.
